modules/azure/client_factory_test.go

```diff
@@ -0,0 +1,143 @@
+// +build azure
+
+// This file contains unit tests for the client factory implementation(s).
+
+package azure
+
+import (
+	"os"
+	"testing"
+
+	autorest "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Local consts for this file only
+const govCloudEnvName = "AzureUSGovernmentCloud"
+const publicCloudEnvName = "AzurePublicCloud"
+const chinaCloudEnvName = "AzureChinaCloud"
+const germanyCloudEnvName = "AzureGermanCloud"
+
+func TestDefaultEnvIsPublicWhenNotSet(t *testing.T) {
+	// save any current env value and restore on exit
+	originalEnv := os.Getenv(AzureEnvironmentEnvName)
+	defer os.Setenv(AzureEnvironmentEnvName, originalEnv)
+
+	// Set env var to missing value
+	os.Setenv(AzureEnvironmentEnvName, "")
+
+	// get the default
+	env := getDefaultEnvironmentName()
+
+	// Make sure it's public cloud
+	assert.Equal(t, autorest.PublicCloud.Name, env)
+}
+
+func TestDefaultEnvSetToGov(t *testing.T) {
+	// save any current env value and restore on exit
+	originalEnv := os.Getenv(AzureEnvironmentEnvName)
+	defer os.Setenv(AzureEnvironmentEnvName, originalEnv)
+
+	// Set env var to gov
+	os.Setenv(AzureEnvironmentEnvName, govCloudEnvName)
+
+	// get the default
+	env := getDefaultEnvironmentName()
+
+	// Make sure it's public cloud
+	assert.Equal(t, autorest.USGovernmentCloud.Name, env)
+}
+
+func TestSubscriptionClientBaseURISetCorrectly(t *testing.T) {
+	var cases = []struct {
+		CaseName        string
+		EnvironmentName string
+		ExpectedBaseURI string
+	}{
+		{"GovCloud/SubscriptionClient", govCloudEnvName, autorest.USGovernmentCloud.ResourceManagerEndpoint},
+		{"PublicCloud/SubscriptionClient", publicCloudEnvName, autorest.PublicCloud.ResourceManagerEndpoint},
+		{"ChinaCloud/SubscriptionClient", chinaCloudEnvName, autorest.ChinaCloud.ResourceManagerEndpoint},
+		{"GermanCloud/SubscriptionClient", germanyCloudEnvName, autorest.GermanCloud.ResourceManagerEndpoint},
+	}
+
+	// save any current env value and restore on exit
+	currentEnv := os.Getenv(AzureEnvironmentEnvName)
+	defer os.Setenv(AzureEnvironmentEnvName, currentEnv)
+
+	for _, tt := range cases {
```

Comment:
I think you need to store a local copy:

```suggestion
	for _, tt := range cases {
		// The following is necessary to make sure testCase's values don't
		// get updated due to concurrency within the scope of t.Run(..) below	
		tt := tt
```

It might not be necessary as you aren't using `t.Parallel()`, but it's always a good sanity check to have in case that ever changes in the future.